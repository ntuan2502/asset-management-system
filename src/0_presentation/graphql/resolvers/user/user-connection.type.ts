import { ObjectType } from '@nestjs/graphql';
import { Paginated } from 'src/shared/dtos/pagination.dto';
import { UserType } from './user.type';

@ObjectType()
export class UserConnection extends Paginated(UserType) {}
