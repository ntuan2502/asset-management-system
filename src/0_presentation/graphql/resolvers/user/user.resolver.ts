import {
  Args,
  Mutation,
  Resolver,
  ObjectType,
  Field,
  ID,
} from '@nestjs/graphql';
import { CommandBus } from '@nestjs/cqrs';
import { CreateUserInput } from 'src/1_application/user/dtos/create-user.input';
import { CreateUserCommand } from 'src/1_application/user/commands/impl/create-user.command';

// Không cần import UserAggregate ở đây nữa

// Định nghĩa UserType như một class độc lập.
// Đây là "hình dạng" dữ liệu mà GraphQL API sẽ trả về.
@ObjectType('User')
export class UserType {
  @Field(() => ID)
  id: string;

  @Field()
  email: string;

  @Field()
  firstName: string;

  @Field()
  lastName: string;
}

@Resolver(() => UserType)
export class UserResolver {
  constructor(private readonly commandBus: CommandBus) {}

  @Mutation(() => UserType)
  async createUser(@Args('input') input: CreateUserInput): Promise<UserType> {
    // CommandBus.execute sẽ trả về một UserAggregate.
    // Vì UserAggregate và UserType có cấu trúc tương tự (duck typing),
    // TypeScript sẽ cho phép điều này.
    // Kết quả trả về cho client sẽ được định dạng theo UserType.
    return this.commandBus.execute(new CreateUserCommand(input));
  }
}
