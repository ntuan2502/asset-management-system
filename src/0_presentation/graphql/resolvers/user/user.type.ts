import { ObjectType, Field, ID, GraphQLISODateTime } from '@nestjs/graphql';

@ObjectType('User')
export class UserType {
  @Field(() => ID)
  id: string;

  @Field()
  email: string;

  @Field()
  firstName: string;

  @Field()
  lastName: string;

  // --- THÊM MỚI CÁC TRƯỜNG ---

  @Field(() => GraphQLISODateTime, { nullable: true }) // Dùng kiểu DateTime của GraphQL
  dob: Date;

  @Field({ nullable: true })
  gender: string;

  @Field(() => GraphQLISODateTime) // Kiểu DateTime cho ngày tháng
  createdAt: Date;

  @Field(() => GraphQLISODateTime) // Kiểu DateTime cho ngày tháng
  updatedAt: Date;
}
