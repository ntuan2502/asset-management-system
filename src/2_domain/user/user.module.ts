import { Module } from '@nestjs/common';
import { CqrsModule } from '@nestjs/cqrs';
import { UserResolver } from 'src/0_presentation/graphql/resolvers/user/user.resolver';
import { CreateUserHandler } from 'src/1_application/user/commands/handlers/create-user.handler';
import { PrismaService } from 'src/3_infrastructure/persistence/prisma/prisma.service';
import { PrismaUserRepository } from 'src/3_infrastructure/persistence/prisma/repositories/prisma-user.repository';
import { USER_REPOSITORY } from './repositories/user.repository.interface';
import { GetUserByIdHandler } from 'src/1_application/user/queries/handlers/get-user-by-id.handler'; // << IMPORT
import { GetAllUsersHandler } from 'src/1_application/user/queries/handlers/get-all-users.handler';

const commandHandlers = [CreateUserHandler];
const queryHandlers = [GetUserByIdHandler, GetAllUsersHandler]; // << THÊM HANDLER VÀO ĐÂY
const repositories = [
  {
    provide: USER_REPOSITORY,
    useClass: PrismaUserRepository,
  },
];

@Module({
  imports: [CqrsModule],
  providers: [
    UserResolver,
    PrismaService,
    ...commandHandlers,
    ...queryHandlers, // << ĐĂNG KÝ
    ...repositories,
  ],
})
export class UserModule {}
