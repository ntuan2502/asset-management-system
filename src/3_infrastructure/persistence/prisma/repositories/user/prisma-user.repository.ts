import { Injectable } from '@nestjs/common';
import {
  IUserRepository,
  PaginatedUsers,
  UserWithPermissions,
} from 'src/2_domain/user/repositories/user.repository.interface';
import { UserAggregate } from 'src/2_domain/user/aggregates/user.aggregate';
import { PrismaService } from '../../prisma.service';
import { UserMapper } from 'src/1_application/user/mappers/user.mapper';

@Injectable()
export class PrismaUserRepository implements IUserRepository {
  constructor(private readonly prisma: PrismaService) {}

  async create(user: UserAggregate): Promise<UserAggregate> {
    const prismaUserData = UserMapper.toPersistence(user);

    const createdPrismaUser = await this.prisma.user.create({
      data: prismaUserData,
    });
    return UserMapper.toDomain(createdPrismaUser);
  }

  async findById(id: string): Promise<UserAggregate | null> {
    const prismaUser = await this.prisma.user.findFirst({
      where: {
        id,
        deletedAt: null,
      },
      include: { roles: true },
    });
    return prismaUser ? UserMapper.toDomain(prismaUser) : null;
  }

  async findByEmail(email: string): Promise<UserAggregate | null> {
    const prismaUser = await this.prisma.user.findFirst({
      where: {
        email,
      },
      include: { roles: true },
    });
    return prismaUser ? UserMapper.toDomain(prismaUser) : null;
  }

  async findAll(args: {
    page: number;
    limit: number;
  }): Promise<PaginatedUsers> {
    const { page, limit } = args;
    const skip = (page - 1) * limit;

    const [users, totalCount] = await this.prisma.$transaction([
      this.prisma.user.findMany({
        where: { deletedAt: null },
        include: { roles: true },
        skip: skip,
        take: limit,
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.user.count({ where: { deletedAt: null } }),
    ]);

    return {
      nodes: users.map((user) => UserMapper.toDomain(user)),
      meta: {
        totalCount,
        page,
        limit,
      },
    };
  }

  async softDelete(id: string): Promise<void> {
    await this.prisma.user.update({
      where: { id },
      data: {
        deletedAt: new Date(),
      },
    });
  }

  async findWithPermissionsByEmail(
    email: string,
  ): Promise<UserWithPermissions | null> {
    const user = await this.prisma.user.findUnique({
      where: { email },
      include: {
        roles: {
          include: {
            permissions: true,
          },
        },
      },
    });

    if (!user) {
      return null;
    }

    const permissions = user.roles.flatMap((role) => role.permissions);
    const uniquePermissions = [
      ...new Map(
        permissions.map((p) => [`${p.action}:${p.subject}`, p]),
      ).values(),
    ];

    const aggregate = UserMapper.toDomain(user);

    (aggregate as UserWithPermissions).permissions = uniquePermissions.map(
      (p) => ({
        action: p.action,
        subject: p.subject,
      }),
    );

    return aggregate as UserWithPermissions;
  }

  async findWithPermissionsById(
    id: string,
  ): Promise<UserWithPermissions | null> {
    const user = await this.prisma.user.findUnique({
      where: { id },
      include: {
        roles: {
          include: {
            permissions: true,
          },
        },
      },
    });

    if (!user) {
      return null;
    }

    const permissions = user.roles.flatMap((role) => role.permissions);
    const uniquePermissions = [
      ...new Map(
        permissions.map((p) => [`${p.action}:${p.subject}`, p]),
      ).values(),
    ];

    const aggregate = UserMapper.toDomain(user);

    (aggregate as UserWithPermissions).permissions = uniquePermissions.map(
      (p) => ({
        action: p.action,
        subject: p.subject,
      }),
    );

    return aggregate as UserWithPermissions;
  }
}
