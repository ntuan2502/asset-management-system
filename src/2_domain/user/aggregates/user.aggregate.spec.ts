import { UserAggregate } from './user.aggregate';
import { UserDeletedEvent } from '../events/user-deleted.event';
import { UserCreatedEvent } from '../events/user-created.event';

// describe: Nhóm các bài test cho UserAggregate
describe('UserAggregate', () => {
  // it: Mô tả một kịch bản test cụ thể
  it('should apply a UserCreatedEvent and update state correctly', () => {
    // --- 1. Sắp đặt (Arrange) ---
    const aggregate = new UserAggregate();
    const userData = {
      id: 'user-123',
      email: 'test@example.com',
      hashedPassword: 'hashed-password',
      firstName: 'John',
      lastName: 'Doe',
      dob: new Date('1990-01-15T00:00:00.000Z'),
      gender: 'MALE',
    };

    jest.useFakeTimers().setSystemTime(new Date('2025-01-01T10:00:00.000Z'));

    // --- 2. Hành động (Act) ---
    aggregate.createUser(userData);

    // --- 3. Khẳng định (Assert) ---
    const uncommittedEvents = aggregate.getUncommittedEvents();
    const createdEvent = uncommittedEvents[0];

    // **Khẳng định tường minh cho những phần quan trọng nhất**
    // "Tôi khẳng định rằng PHẢI có đúng 1 sự kiện được phát ra"
    expect(uncommittedEvents).toHaveLength(1);
    // "Và sự kiện đó PHẢI là UserCreatedEvent"
    expect(createdEvent).toBeInstanceOf(UserCreatedEvent);
    // "Và ID của nó PHẢI khớp với ID tôi đã cung cấp"
    expect((createdEvent as UserCreatedEvent).id).toBe(userData.id);

    // **Dùng Snapshot để kiểm tra phần còn lại**
    // "Bây giờ, hãy chụp ảnh toàn bộ sự kiện để đảm bảo các trường khác
    // (firstName, lastName, dob...) không bị thay đổi ngoài ý muốn."
    expect(createdEvent).toMatchSnapshot();

    // Tương tự, kiểm tra trạng thái của aggregate
    // "Tôi khẳng định rằng ID và Email của aggregate PHẢI được cập nhật đúng"
    expect(aggregate.id).toBe(userData.id);
    expect(aggregate.email).toBe(userData.email);
    // "Và bây giờ, chụp ảnh phần còn lại của aggregate"
    expect(aggregate).toMatchSnapshot();

    jest.useRealTimers();
  });

  it('should apply UserDeletedEvent when deleting an active user', () => {
    // 1. Arrange
    // Tạo một user đã được khởi tạo (giống như đã có UserCreatedEvent)
    // để có ID và trạng thái active.
    const aggregate = new UserAggregate();
    aggregate.id = 'user-123';
    aggregate.deletedAt = null; // Tường minh hóa trạng thái active

    // 2. Act
    aggregate.deleteUser();

    // 3. Assert
    const uncommittedEvents = aggregate.getUncommittedEvents();

    // Khẳng định rằng có một sự kiện được phát ra
    expect(uncommittedEvents).toHaveLength(1);

    // Khẳng định rằng đó là đúng sự kiện UserDeletedEvent
    const event = uncommittedEvents[0];
    expect(event).toBeInstanceOf(UserDeletedEvent);
    expect((event as UserDeletedEvent).id).toBe('user-123');

    // (Tùy chọn) Khẳng định rằng trạng thái nội bộ của aggregate đã được cập nhật
    expect(aggregate.deletedAt).not.toBeNull();
  });

  it('should throw an error when trying to delete an already deleted user', () => {
    // --- 1. Sắp đặt (Arrange) ---
    // Tạo một user đã ở trạng thái bị xóa
    const aggregate = new UserAggregate();
    // Giả lập trạng thái đã bị xóa
    aggregate.deletedAt = new Date();

    // --- 2. Hành động & 3. Khẳng định (Act & Assert) ---
    // Chúng ta mong đợi rằng khi gọi `deleteUser`, một lỗi sẽ được ném ra.
    // Chúng ta bọc hành động trong một arrow function.
    const action = () => aggregate.deleteUser();

    // `toThrow` kiểm tra xem hàm có ném ra lỗi không.
    // Bạn cũng có thể kiểm tra nội dung của thông báo lỗi.
    expect(action).toThrow(
      'Cannot delete a user that has already been deleted.',
    );
  });

  it('should throw an error when trying to restore an already active user', () => {
    const aggregate = new UserAggregate();
    const action = () => aggregate.restoreUser();
    expect(action).toThrow('Cannot restore an active user.');
  });
});
