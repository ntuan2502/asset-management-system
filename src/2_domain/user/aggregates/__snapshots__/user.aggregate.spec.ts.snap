// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAggregate should apply a UserCreatedEvent and update state correctly 1`] = `
UserCreatedEvent {
  "createdAt": 2025-01-01T10:00:00.000Z,
  "departmentId": "dept-xyz",
  "dob": 1990-01-15T00:00:00.000Z,
  "email": "test@example.com",
  "firstName": "John",
  "gender": "MALE",
  "hashedPassword": "hashed-password",
  "id": "user-123",
  "lastName": "Doe",
  "officeId": "office-abc",
}
`;

exports[`UserAggregate should apply a UserCreatedEvent and update state correctly 2`] = `
UserAggregate {
  "aggregateType": "User",
  "createdAt": 2025-01-01T10:00:00.000Z,
  "deletedAt": null,
  "departmentId": "dept-xyz",
  "dob": 1990-01-15T00:00:00.000Z,
  "email": "test@example.com",
  "firstName": "John",
  "gender": "MALE",
  "id": "user-123",
  "lastName": "Doe",
  "officeId": "office-abc",
  "password": "hashed-password",
  "roleIds": [],
  "updatedAt": 2025-01-01T10:00:00.000Z,
  "version": 0,
  Symbol(): false,
  Symbol(): [
    UserCreatedEvent {
      "createdAt": 2025-01-01T10:00:00.000Z,
      "departmentId": "dept-xyz",
      "dob": 1990-01-15T00:00:00.000Z,
      "email": "test@example.com",
      "firstName": "John",
      "gender": "MALE",
      "hashedPassword": "hashed-password",
      "id": "user-123",
      "lastName": "Doe",
      "officeId": "office-abc",
    },
  ],
}
`;
